components/layout/display_list_builder.rs

```diff
@@ -1184,172 +1367,100 @@ impl FragmentDisplayListBuilding for Fragment {
+            ),
+        };
 
-            (BackgroundSize::Explicit { width, height: LengthOrPercentageOrAuto::Auto }, _) => {
-                let width = MaybeAuto::from_style(width, bounds.size.width)
-                                      .specified_or_default(intrinsic_size.width);
-                Size2D::new(width, Au::from_f64_px(width.to_f64_px() / image_aspect_ratio))
-            }
+        let mut bounds = match bg_attachment {
+            BackgroundAttachment::Scroll => match bg_origin {
+                BackgroundOrigin::BorderBox => absolute_bounds,
+                BackgroundOrigin::PaddingBox => calculate_inner_bounds(
+                    absolute_bounds,
+                    style.logical_border_width().to_physical(style.writing_mode),
+                ),
+                BackgroundOrigin::ContentBox => calculate_inner_bounds(
+                    absolute_bounds,
+                    self.border_padding.to_physical(style.writing_mode),
+                ),
+            },
+            BackgroundAttachment::Fixed => Rect::new(
+                Point2D::origin(),
+                // Get current viewport
+                state.layout_context.shared_context().viewport_size(),
+            ),
+        };
 
-            (BackgroundSize::Explicit { width: LengthOrPercentageOrAuto::Auto, height }, _) => {
-                let height = MaybeAuto::from_style(height, bounds.size.height)
-                                       .specified_or_default(intrinsic_size.height);
-                Size2D::new(Au::from_f64_px(height.to_f64_px() * image_aspect_ratio), height)
-            }
+        let mut tile_size = compute_background_image_size(bg_size, bounds.size, intrinsic_size);
 
-            (BackgroundSize::Explicit { width, height }, _) => {
-                Size2D::new(MaybeAuto::from_style(width, bounds.size.width)
-                                 .specified_or_default(intrinsic_size.width),
-                       MaybeAuto::from_style(height, bounds.size.height)
-                                 .specified_or_default(intrinsic_size.height))
-            }
-        }
+        let mut tile_spacing = Size2D::zero();
+        let own_position = bounds.size - intrinsic_size.unwrap_or(Size2D::zero());
+        let pos_x = bg_position_x.to_used_value(own_position.width);
+        let pos_y = bg_position_y.to_used_value(own_position.height);
+        tile_image_axis(
+            bg_repeat.0,
+            &mut bounds.origin.x,
+            &mut bounds.size.width,
+            &mut tile_size.width,
+            &mut tile_spacing.width,
+            pos_x,
+            css_clip.origin.x,
+            css_clip.size.width);
+        tile_image_axis(
+            bg_repeat.1,
+            &mut bounds.origin.y,
+            &mut bounds.size.height,
+            &mut tile_size.height,
+            &mut tile_spacing.height,
+            pos_y,
+            css_clip.origin.y,
+            css_clip.size.height);
```

Comment:
nit: Paren on the next line, here and in the rest of the calls using this format.